pkg/raft/raft.go

```diff
@@ -1040,6 +1034,51 @@ func (r *raft) reset(term uint64) {
 	r.uncommittedSize = 0
 }
 
+func (r *raft) setTerm(term uint64) {
+	if term == r.Term {
+		return
+	}
+	assertTrue(term > r.Term, "term cannot regress")
+	r.Term = term
+	r.Vote = None
+	r.lead = None
+	r.leadEpoch = 0
+}
+
+func (r *raft) setVote(id pb.PeerID) {
+	if id == r.Vote {
+		return
+	}
+	assertTrue(r.Vote == None, "cannot change vote")
+	r.Vote = id
+}
+
+func (r *raft) setLead(lead pb.PeerID) {
+	if lead == r.lead {
+		return
+	}
+	assertTrue(r.lead == None, "cannot change lead")
+	r.lead = lead
+}
+
+func (r *raft) resetLead() {
+	r.lead = None
+	r.leadEpoch = 0
+}
+
+func (r *raft) setLeadEpoch(leadEpoch pb.Epoch) {
+	if leadEpoch == r.leadEpoch {
+		return
+	}
+	assertTrue(r.lead != None, "leader must be set")
+	assertTrue(leadEpoch > r.leadEpoch, "leadEpoch cannot regress")
+	r.leadEpoch = leadEpoch
+}
+
+func (r *raft) resetLeadEpoch() {
+	r.leadEpoch = 0
```

Comment:
Non-actionable, but just curious about the fortification protocol: is it safe to reset `leadEpoch` to 0 in the de-fortify case?

Can the `leadEpoch` be set to non-zero again during the course of the term? If yes, can it be set to a value below the last non-zero `leadEpoch`?

Reply:
It is safe to reset to 0 when if either (1) the leader has told us to defortify, or (2) we have withdrawn support for the lead epoch in the store liveness network.

`leadEpoch` can then be set to a non-zero value again if a new `MsgForgetLeader` is received, which is by design. However, the value must be equal or greater than it was before it was reset to 0, because it comes from a call to `storeLiveness.SupportFor`, which is guaranteed to be monotonic.